### ISSUE TITLE
Improve memory allocations of DeflateStream

### ISSUE THREAD

**Issue by redknightlois** [dotnet/corefx#1991]
We use GZipStream to handle the data sent on the wire to a WebAPI with both big and small requests. The allocation stress is more evident with small packages though. 

On a sample of 60 seconds in an stress test we measured that
System.IO.Compression.OutputWindow.ctor() and System.IO.Compression.DefrateStream.ctor() would be consistently in the top 5 allocators. Between the two 1.8Gb in allocations in 60 seconds.

The root is that DefrateStream will create on every construction (1 per request) an Infrater (https://github.com/dotnet/corefx/blob/master/src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs#L62) and the Infrater will also build an OutputWindow (https://github.com/dotnet/corefx/blob/master/src/System.IO.Compression/src/System/IO/Compression/OutputWindow.cs). Which in turn will allocate 32Kb to use as a buffer for each request. 

Reusing those buffers and/or OutputWindow would aleviate a lot the allocation stress GZipStream causes when used in small streams contexts like Web APIs.

**Comment by stephentoub** [comment]
Could you share some more details? e.g. how many requests were there that represented those allocations?  How many were running concurrently on average and at the limit? (And what do you expect in a normal use case vs stress?) etc.  I ask because any pooling strategy would need to be tuned appropriately.

**Comment by redknightlois** [comment]
Sure, I can even send you the code to repro, or the profiler session dump if you want it. We were analyzing our own hotspots when we stumbled upon this (completely unintented finding), we usually just go and fix them (sometimes using another library or writing our own components if we need to) but wanted to report it so you can have it in your radar as the current API does not provide a way to fix it at all. 

The test is synthetic so message size and some other parameters are controlled for what we wanted to break, but it roughly represents a typical heavy write use case scenario (small documents puts, deletes and updates at high frequency). 
Code: https://gist.github.com/redknightlois/b0f38b4fac93af4fd137

Some details:

- Test time: Aprox 60 seconds (under the profiler so those may be a bit less in a real scenario)
- Mean Message Size (including HTTP headers): 300 bytes
- Number of requests: 52486 request (compress & decompress)
- Number of decompress requests: 26243 requests (aka WindowOutput.ctor calls)
- Number of client threads: 14 (initiators)
- Number of handling threads: 8 (server managed)
- Total memory allocated: 17Gb 
- Total memory allocated by GZipStream: 1.8Gb (roughly 10% of total -- but we have found some of our code that needs fixing, so this percentage is going to go up real soon).

While the user may need to know what he is doing, having a way to provide your own buffer would not be such a bad idea either.

**Comment by stephentoub** [comment]
Thanks for the extra info.  I'm marking this as up-for-grabs in case anyone wants to take a look before we get around to investigating.  

Rather than having GZipStream/DeflateStream try to manage a pool of buffes and the like, I'm wondering whether it instead makes more sense to allow the consumer to essentially reuse a GZipStream/DeflateStream by resetting it to wrap a different underlying Stream; then the consuming applications can manage the pooling however they see fit.

**Comment by redknightlois** [comment]
That would have been my first option if changing the semantics were an allowed change.

**Comment by friedroman** [comment]
I also encountered this issue during profiling my application. Not http but websocket compression transferring around 70 Mb uncompressed / 17 Mb compressed JSON.
In my case during 1 min profiling session 1024 Mb out of total 4788 was allocated by GZipStream. Each websocket message is compressed individually therefore I have to instantiate new GZipStream for each message.
Ability to reuse the stream @stephentoub suggested was the first thing I looked for (fruitless, of course). But I would not have started looking for it if I didn't know about the problem. Not a Pit of Success story even if it was reusable.

**Comment by redknightlois** [comment]
@stephentoub Got some time to look into this while waiting until another issue which is blocking me. I am having a hard time to build in VS2015 CoreFX, if I manage to make it work I plan to add the .Reset( Stream stream ) to GZipStream/DefrateStream. However, that is a public API Change, should be done on the Future branch? Is there any formal API review work we should do before starting?

**Comment by stephentoub** [comment]
@redknightlois, thanks.  The API review process is outlined at https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md.

**Comment by redknightlois** [comment]
```DefrateStream``` allocates on average 40Kb per instance just in usage buffers. When those Streams are used to encode short messages like HTTP messages in WebApi, a significant allocation stress happens on the GC. A 100 request/sec WebApi (very low throughput) using gzipped content will cause 236Mb of memory allocated per minute. 

On a real scenario:

- Test time: Aprox 60 seconds (under the profiler so those may be a bit less in a real scenario)
- Mean Message Size (including HTTP headers): 300 bytes
- Number of requests: 52486 request (compress & decompress)
- Number of decompress requests: 26243 requests (aka WindowOutput.ctor calls)
- Number of client threads: 14 (initiators)
- Number of handling threads: 8 (server managed)
- Total memory allocated: 17Gb
- Total memory allocated by GZipStream: 1.8Gb (roughly 10% of total -- after fixing this is roughly 30% of the allocations).

### Proposed Solutions ###

#### Using ObjectPool<T> to allocate internal buffers ####

The implementation based on ```ObjectPool<T>``` will not require any change to the API and/or the code used by the user of either ```DefrateStream``` and ```GZipStream```. This solution aims to provide a plug-and-play improvement on the allocation behavior. The cons is that selecting the appropriate size of total memory to "allocate" is error prone and wont work for every workload. 

The PR presented for review has selected the magic numbers based on the following rationale:

- Keep the total committed memory under 1MB in the worst case scenario
- Assuming a 50-50 distribution between decompress/compress tasks.
- Assumes that compression/decompression time is essentially small.
- ```Streams``` are immediately disposed upon usage. 

The behavior for this solution if any of those pre-requisites are is not uphold is the current behavior. 

PR for consideration: https://github.com/dotnet/corefx/pull/3702

#### Allow sharing instances between multiple incoming connections ####

This implementation requires changing the public API and also the users' code to gain performance. The base idea is that a ```GZipStream / DefrateStream``` state can be resetted with a new input stream, allowing to cache and reuse it instances. 

Additions to the current public API:

```csharp
public class DefrateStream
{
   public bool IsOpen { get; }
   public virtual void Close();
   public virtual void Reset (Stream stream);
}

public class GZipStream
{
   public bool IsOpen { get; }
   public virtual void Close();
   public virtual void Reset (Stream stream);
}
```

While this mechanism allows the implementers to accomodate the use of ```DefrateStream``` and ```GZipStream``` to their workload / memory pressure environment, current users will not get any benefic. Furthermore, currently the implementation assumes that a Stream will finish writing to the output stream only when it is disposed, which is incompatible with this approach. To account for that a ```.Close()``` method and a ```.Reset()``` method had to be added and rules of usage had to be amended, which is a bit misleading. 

PR for consideration: https://github.com/dotnet/corefx/pull/3703

### Conclusions ###

My feeling is that this approach will require a completely reworked base API under the name ```ReusableGZipStream``` and ```ReusableDefrateStream``` with a very strict but different usage model (which probably would not inherit from Stream at all). Nonetheless, I am making the current implementation available for review in another PR for reference purposes.

**Comment by ianhays** [comment]
Closing this as we no longer use the managed Deflate implementation, so reusing OuterWindow instances is a non-issue.
